### ISSUE TITLE
Bug: no-unnecessary-condition False positive erroring on valid code with incorrect message

### ISSUE THREAD

**Issue by web-dev-sam** [typescript-eslint/typescript-eslint#11856]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.9.2&fileType=.tsx&code=FAYw9gdgzgLgBAUwDYILYIjKcC8cAmYIArupgHQCOxCATgJ4DKyCIMYtAgkkgBQDkAQwAWAKyj5h-AJTAU8AJZQAwkgUgA1gnwBJaAvwJccAGaCkUBMGAsyWciY4BRQSGG9etjDGm4AfHAA3sBwcAomcJ4oduTgmIIK0LwQCADucAByYIa80nlBcAD0hXAASggADkiuRqkKMMJwtAjmmdkIADRwUILoYVBQNCGh-arqWrr6hsYwtEOhAL7AC7LA4ZEAhEpjmtp6UAYIvoFFJQAqwkqItLQc-WEQ4Des8AePRttqu5MH0yCCEDgACMjLN5nA4lAwChyEgwABzAQAPTgAFUICkQAgBoIGBDIPh6gpIOYugA3cw0e7mVKCejYMENejkGTLYBAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
const elements = document.querySelectorAll('ahjsdh')
let isClickedInside = false

elements.forEach((element) => {
  if (element.contains(new Node())) { // Replace with real Node, same issue
    isClickedInside = true
  }
})

if (!isClickedInside) { // This error is incorrect since isClickedInside can be true
  console.log('^ Unnecessary conditional, value is always truthy.')
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unnecessary-condition": "error"
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {}
}
```

### Expected Result

I expected no error to occur.

### Actual Result

There was an error this line: `if (!isClickedInside) { `

**Comment by kirkwaiblinger** [comment]
Hi @web-dev-sam!

This is a [classic limitation of the rule](https://typescript-eslint.io/rules/no-unnecessary-condition/#values-modified-within-function-calls). TS doesn't have a way to indicate that functions may a) be executed immediately or b) need to reset narrowings of captured variables. Therefore, according to the types, the rule is behaving correctly.

Note that [a `for...of` loop does not suffer this limitation](https://typescript-eslint.io/play/#ts=5.9.2&fileType=.tsx&code=MYewdgzgLgBApgGzgWzmKEYF4YBMTACuq6AdAI6FwBOAngMqJzBQjUCCCCAFAOQCGACwBWEXIN4BKAFBJYASwgBhBPOABrOLgCSkebjjYYAM34IIcadONsY3UJFhMSsEMfhIXESTADe0mBh5d25nNChSByh%2BeUhuMDgAdxgAORADbkksvxgAelyYACU4AAcEfmBDRPkoQRhqODNU9LgAGhgIflQgiAgqAMCelTVNHT0DIyhqfsCAX2l56WC7AEJFYY0tXQh9OB9fPIKAFUFFeGpqWzPY0AvmWB2wSqHVTbGdieB%2BMBgAI0MpjMYA4ICAkKQECAAOZ8AB6MAAqmAEpVevw6MDwLgavJwGZ2gA3MxUHowMyJfi0TCA2q0UhSBZAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false), which is one of the many reasons I prefer to use `for...of` whenever possible.
